presto-hive/src/main/java/com/facebook/presto/hive/IonSqlQueryBuilder.java

```diff
@@ -0,0 +1,297 @@
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.DAYS;
+import static java.util.stream.Collectors.joining;
+import static org.joda.time.chrono.ISOChronology.getInstanceUTC;
+import static org.joda.time.format.ISODateTimeFormat.date;
+
+/**
+ * S3 Select uses Ion SQL++ query language. This class is used to construct a valid Ion SQL++ query
+ * to be evaluated with S3 Select on an S3 object.
+ */
+public class IonSqlQueryBuilder
+{
+    private static final DateTimeFormatter FORMATTER = date().withChronology(getInstanceUTC());
+    private static final String DATA_SOURCE = "S3Object s";
+    private final TypeManager typeManager;
+
+    IonSqlQueryBuilder(TypeManager typeManager)
+    {
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+    }
+
+    String buildSql(List<HiveColumnHandle> columns, TupleDomain<HiveColumnHandle> tupleDomain)
+    {
+        StringBuilder sql = new StringBuilder("SELECT ");
+
+        if (columns.isEmpty()) {
+            sql.append("' '");
+        }
+        else {
+            String columnNames = columns.stream()
+                    .map(column -> format("s._%d", column.getHiveColumnIndex() + 1))
+                    .collect(joining(", "));
+            sql.append(columnNames);
+        }
+
+        sql.append(" FROM ");
+        sql.append(DATA_SOURCE);
+
+        List<String> clauses = toConjuncts(columns, tupleDomain);
+        if (!clauses.isEmpty()) {
+            sql.append(" WHERE ")
+                    .append(Joiner.on(" AND ").join(clauses));
+        }
+
+        return sql.toString();
+    }
+
+    private List<String> toConjuncts(List<HiveColumnHandle> columns, TupleDomain<HiveColumnHandle> tupleDomain)
+    {
+        Builder<String> builder = builder();
+        for (HiveColumnHandle column : columns) {
+            Type type = column.getHiveType().getType(typeManager);
+            if (tupleDomain.getDomains().isPresent() && isSupported(type)) {
+                Domain domain = tupleDomain.getDomains().get().get(column);
+                if (domain != null) {
+                    builder.add(toPredicate(domain, type, column.getHiveColumnIndex()));
+                }
+            }
+        }
+        return builder.build();
+    }
+
+    private static boolean isSupported(Type type)
+    {
+        Type validType = requireNonNull(type, "type is null");
+        return validType.equals(BIGINT) ||
+                validType.equals(TINYINT) ||
+                validType.equals(SMALLINT) ||
+                validType.equals(INTEGER) ||
+                validType instanceof DecimalType ||
+                /*
+                 * Double and Real Types loose precision. Thus, they are not pushed down to S3. Please use Decimal Type if push down is desired.
```

Comment:
Nit: "lose" rather than "loose"

Since these types are not supported, should they be removed from `valueToQuery()` below and the `SUPPORTED_COLUMN_TYPES` list?

Reply:
We noticed issues with supporting these types during our early integration testing. Thus, we commented this code out with a reason for why they are not supported. 

It's essentially dead code. I think it's better to remove it for now and we can add it back later. I will move the comments to higher level.